cpp/ql/test/experimental/query-tests/Security/CWE/CWE-193/pointer-deref/InvalidPointerToDereference.ql

```diff
@@ -0,0 +1,81 @@
+import cpp
+import semmle.code.cpp.security.InvalidPointerDereference.InvalidPointerToDereference
+import TestUtilities.InlineExpectationsTest
+import semmle.code.cpp.ir.IR
+import semmle.code.cpp.dataflow.new.DataFlow
+
+string case3(DataFlow::Node derefSource, DataFlow::Node derefSink, DataFlow::Node operation) {
+  operationIsOffBy(_, _, derefSource, derefSink, _, operation, _) and
+  not exists(case2(_, _, operation)) and
+  not exists(case1(_, _, operation)) and
+  exists(int derefSourceLine, int derefSinkLine, int operationLine |
+    derefSourceLine = derefSource.getLocation().getStartLine() and
+    derefSinkLine = derefSink.getLocation().getStartLine() and
+    operationLine = operation.getLocation().getStartLine() and
+    derefSourceLine != derefSinkLine and
+    derefSinkLine != operationLine and
+    result = "L" + derefSourceLine + "->L" + derefSinkLine + "->L" + operationLine
+  )
+}
+
+string case2(DataFlow::Node derefSource, DataFlow::Node derefSink, DataFlow::Node operation) {
+  operationIsOffBy(_, _, derefSource, derefSink, _, operation, _) and
+  not exists(case1(_, _, operation)) and
```

Comment:
Is this necessary? My initial impression is that this should be covered by:
```ql
    derefSourceLine = derefSinkLine and
    derefSinkLine != operationLine and
```
and similar in `case3`

If the answer is that it's not necessary, I'm fine with leaving this in though, just because it might be slightly clearer to have it.

Reply:
Looking at the test results when I remove the lines it looks like it's necessary. I think the explanation is that there may be several `derefSource -> derefSink -> operation` pairs that lead to the same `operation`. Something like:
```cpp
4. char* p = malloc(size);
5. char* q = p + size; // $ alloc=L4
6. char a = *q; // $ deref=L6 // BAD
```
Without the restriction, we need the annotation `deref=L5->L6` because range analysis identifies both `p + size` on line 5, and `q` on line 6 as the `derefSource` (because they have identical value numbers). But when we include the `exists(caseX(_, _, operation))` cases we ensure that there's not a "smaller" path that gets us to the operation.

Reply:
Tanks for checking!